app/main/views/agreements.py

```diff
@@ -77,3 +148,54 @@ def update_framework_agreement(agreement_id):
         return jsonify(message="Database Error: {0}".format(e)), 400
 
     return jsonify(agreement=framework_agreement.serialize())
+
+
+@main.route('/agreements/<int:agreement_id>/sign', methods=['POST'])
+def sign_framework_agreement(agreement_id):
+    framework_agreement = FrameworkAgreement.query.filter(FrameworkAgreement.id == agreement_id).first_or_404()
+    framework_agreement_details = framework_agreement.supplier_framework.framework.framework_agreement_details
+
+    json_payload = get_json_from_request()
+    updater_json = validate_and_return_updater_request()
+    update_json = None
+
+    if framework_agreement_details and framework_agreement_details.get('frameworkAgreementVersion'):
+        json_has_required_keys(json_payload, ["agreement"])
+        update_json = json_payload["agreement"]
+
+        json_has_keys(update_json, required_keys=['signedAgreementDetails'])
+
+        framework_agreement.update_signed_agreement_details_from_json(update_json['signedAgreementDetails'])
+        framework_agreement.update_signed_agreement_details_from_json(
+            {'frameworkAgreementVersion': framework_agreement_details['frameworkAgreementVersion']}
+        )
+        validate_agreement_details_data(
+            framework_agreement.signed_agreement_details,
+            enforce_required=True
+        )
+
+        user = User.query.filter(User.id == update_json['signedAgreementDetails']['uploaderUserId']).first()
```

Comment:
This `user` isn't used again after this. Not sure if it needs to be?  If not then don't bother assigning it, just check it exists.